packages/lodestar/src/chain/blocks/process.ts

```diff
@@ -126,27 +131,43 @@ export function updateForkChoice(
 }
 
 export async function runStateTransition(
-  config: IBeaconConfig,
-  db: IBeaconDb,
-  logger: ILogger,
-  stateContext: Required<IStateContext>,
+  emitter: ChainEventEmitter,
+  stateContext: ITreeStateContext,
   job: IBlockProcessJob
 ): Promise<IStateContext | null> {
+  const config = stateContext.epochCtx.config;
   try {
+    const preSlot = stateContext.state.slot;
+    const postSlot = job.signedBlock.message.slot;
+    const preEpoch = computeEpochAtSlot(config, preSlot);
+    let nextEpochSlot = computeStartSlotAtEpoch(config, preEpoch + 1);
+    while (nextEpochSlot < postSlot) {
+      processSlots(stateContext.epochCtx, stateContext.state, nextEpochSlot);
+      const checkpoint: Checkpoint = {
+        root: stateContext.state.blockRoots[(nextEpochSlot - 1) % config.params.SLOTS_PER_HISTORICAL_ROOT],
+        epoch: computeEpochAtSlot(config, nextEpochSlot),
+      };
+      emitter.emit("checkpoint", checkpoint, stateContext);
+      nextEpochSlot = nextEpochSlot + config.params.SLOTS_PER_EPOCH;
+    }
     // if block is trusted don't verify proposer or op signature
-    return fastStateTransition(stateContext, job.signedBlock, {
+    const postStateContext = fastStateTransition(stateContext, job.signedBlock, {
       verifyStateRoot: true,
       verifyProposer: !job.trusted,
       verifySignatures: !job.trusted,
-    });
+    }) as ITreeStateContext;
+    const blockSlot = job.signedBlock.message.slot;
+    if (blockSlot % config.params.SLOTS_PER_EPOCH === 0) {
+      const checkpoint: Checkpoint = {
+        root: stateContext.state.blockRoots[Math.max(0, blockSlot - 1) % config.params.SLOTS_PER_HISTORICAL_ROOT],
```

Comment:
I don't get why it is `blockSlot - 1` instead of `blockSlot`. Should we use `getBlockRootAtSlot` instead?